### DISCUSSION TITLE
Multiple instances of the same Service, but different configuration files ?

### DISCUSSION THREAD

**Discussion by Montago** [CoreWCF/CoreWCF#1269] [Q&A] [answered]
I'm working on an internal IT Support tool that has access to all 3 environments: Develop, Staging and Release. 

The WPF "Frontend" needs to be able to switch between each envirment on the fly (this is already supported and working) 

I need to somehow host all 3 enviroments on the same backend server and have multiple options to choose from:

1. Three different website instances in IIS
    **Negative**: Requires that i deploy the service 3 times with each Build Configuration .. this is uncool 
 
2. Send the environment information in each service call (This is how we did it in .Net Framework 4.8 ... ) And i've so far replicated this feature by adding a custom header to the WCF Message using Message Inspector... 
 **Uknown**: How do i dynamically change the DbContext's based on the header ? .. the attempts i've made so far hasn't been pretty and causes all kinds of problems with DbReaders etc... 
 There's no easy way to do this when it seems like the WCFService instance is not activated for each Service Call ... 
 
 3. Create 3 instances of the service on the same server : 
 - /Service_dev.svc
 - /Service_staging.svc 
 - /Service_prod.svc 
 - Each instance should load their DBContext using the 3 different appsettings.json
 - When WPF Client selects environment, i simply change the active WCFClient ... 
 **Unsure** : How to accomplish this on the Backend !

Could someone tell me if this option 3 is possible ? 
and pretty please, give me an example of how the backend initializer would look like :)

**Comment by Montago** [comment, marked as answer]
So .. I've made succes by created 3 inherited class instances of the Service , each called MyService_Development, MyService_Staging etc.. 

Each invokes a new instance of the Base using each Enviroment and fetches its own DBContext ... 

```
services.AddScoped<MyService_Development>();
services.AddScoped<MyService_Staging>();
services.AddScoped<MyService_Production>();
......
app.UseServiceModel(builder =>
{
    builder.AddService<MyService_Development>();
    builder.AddService<MyService_Staging>();
    builder.AddService<MyService_Production>();
    
    var binding = new BasicHttpBinding(BasicHttpSecurityMode.Transport)
    
    builder.AddServiceEndpoint<MyService_Development, IMyService>(binding, "/MyService_Development.svc", config =>
    {
        config.EndpointBehaviors.Add(new CustomInspectorBehavior());
    });

    builder.AddServiceEndpoint<MyService_Staging, IMyService>(binding, "/MyService_Staging.svc", config =>
    {
        config.EndpointBehaviors.Add(new CustomInspectorBehavior());
    });

    builder.AddServiceEndpoint<MyService_Production, IMyService>(binding, "/MyService_Production.svc", config =>
    {
        config.EndpointBehaviors.Add(new CustomInspectorBehavior());
    });

    serviceMetadataBehavior.HttpsGetEnabled = true;

```

And it seems to work

**Reply by mconnew** [reply]
Another option is to use a custom `IInstanceProvider` and in the `GetInstance(InstanceContext instanceContext, Message message)` examine the `To` header from the message (`Message.Headers.To`) to see which endpoint it was sent to and provide a service instance wired up the way you want it. Le tme know if you want to try that way because I need to check that we always replace the To header with the Uri that the request came in to. With HTTP, the To header isn't included in the SOAP message as it's provided by the Http request Url. My only concern is if the client emits a To header in the SOAP message when it's not needed, that we do overwrite it with the request Url. If we don't do that, then my suggested approach would be vulnerable to a client forcing the request to a different stage. I suspect we do always replace it, but I would need to check before you rely on it.

**Reply by Montago** [reply]
I would love to see an example of how that's all wired up.

And if its easier or harder to read than the solution i came up with.

I'd like to 'do it there right way' whatever that might be. But most important the way which is easy to read for the next unlucky person to take over the code 😅

**Reply by mconnew** [reply]
I threw together a POC that you can find [here](https://github.com/mconnew/Playground/tree/main/ConfigureServiceByEndpoint). It's got a bit more functionality in there than you strictly need, but I maintained the ability to use the `Injected` attribute on operations. Without the extra extension stuff you'll see in the instance provider, that stuff won't work any more. There's only a single method you need to modify, `ConfigureByEndpointInstanceProvider.GetServiceInstanceByPath`. It's just got a switch block to configure the service instance. One thing to keep in mind is this requires PerSession or PerCall instance context mode (PerSession is the default). If you set it to Singleton, you'll get a single instance created and that's what will get used everywhere.
